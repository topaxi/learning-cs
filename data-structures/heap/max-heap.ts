import { Heap } from './heap'
import { neg } from '../../utils/function/neg'

export class MaxHeap<T = number> extends Heap<T> {
  constructor(comparator = (a: T, b: T): number => (a as any) - (b as any)) {
    super(neg(comparator))
  }
}
