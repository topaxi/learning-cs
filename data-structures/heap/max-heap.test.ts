import { MaxHeap } from './max-heap'

describe('MaxHeap', () => {
  test('should create empty heap', () => {
    let heap = new MaxHeap()

    expect(heap.peek()).toBeNull()
    expect(heap.empty).toBe(true)
  })

  test('should push value to heap', () => {
    let heap = new MaxHeap()

    heap.push(42)

    expect(heap.peek()).toBe(42)
    expect(heap.empty).toBe(false)
  })

  test('should push values to heap', () => {
    let heap = new MaxHeap()

    heap.push(43)
    heap.push(44)
    heap.push(46)
    heap.push(42)

    expect(heap.peek()).toBe(46)
    expect(heap.empty).toBe(false)
  })
})
