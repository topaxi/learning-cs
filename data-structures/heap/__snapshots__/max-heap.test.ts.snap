// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaxHeap should push values to heap 1`] = `
Array [
  46,
  44,
  43,
  42,
]
`;
